File: healthcheck.py
import requests
import sys
import os
import time

def check_health(url, max_retries=5, retry_delay=2):
    """
    Check if the application is healthy by making a request to the health endpoint.
    Returns True if healthy, False otherwise.
    """
    for attempt in range(max_retries):
        try:
            response = requests.get(f"{url}/_health", timeout=5)
            if response.status_code == 200:
                print(f"Health check passed on attempt {attempt + 1}")
                return True
            else:
                print(f"Health check failed with status code {response.status_code}")
        except Exception as e:
            print(f"Health check attempt {attempt + 1} failed: {str(e)}")
        
        if attempt < max_retries - 1:
            print(f"Retrying in {retry_delay} seconds...")
            time.sleep(retry_delay)
    
    return False

if __name__ == "__main__":
    # Get the URL from environment or use default
    base_url = os.environ.get("HEALTH_CHECK_URL", "http://localhost:8000")
    
    # Check health and exit with appropriate code
    if check_health(base_url):
        sys.exit(0)  # Success
    else:
        sys.exit(1)  # Failure

File: healthcheck.js
#!/usr/bin/env node

/**
 * Health check script for Railway deployment
 * 
 * This script makes a request to the health check endpoint and exits with
 * the appropriate status code to indicate success or failure.
 */

const http = require('http');
const https = require('https');

// Configuration - use hardcoded port 3000 to match direct_start.js
const PORT = 3000; // Hardcoded to match our application port
const HOST = process.env.HOST || 'localhost';
const PATH = '/_health';
const MAX_RETRIES = 5; // Increase retries
const RETRY_DELAY = 5000; // Increase delay to 5 seconds

// Function to make the health check request
function makeRequest(retry = 0) {
  console.log(`Health check attempt ${retry + 1}/${MAX_RETRIES}...`);
  
  const options = {
    hostname: HOST,
    port: PORT,
    path: PATH,
    method: 'GET',
    timeout: 5000, // 5 second timeout
  };

  const req = http.request(options, (res) => {
    let data = '';
    
    res.on('data', (chunk) => {
      data += chunk;
    });
    
    res.on('end', () => {
      if (res.statusCode === 200) {
        console.log(`Health check passed: ${data}`);
        process.exit(0); // Success
      } else {
        console.error(`Health check failed: Status code ${res.statusCode}`);
        retryOrFail(retry);
      }
    });
  });
  
  req.on('error', (error) => {
    console.error(`Health check error: ${error.message}`);
    retryOrFail(retry);
  });
  
  req.on('timeout', () => {
    console.error('Health check timed out');
    req.destroy();
    retryOrFail(retry);
  });
  
  req.end();
}

// Function to retry or exit with failure
function retryOrFail(currentRetry) {
  if (currentRetry < MAX_RETRIES - 1) {
    console.log(`Retrying in ${RETRY_DELAY / 1000} seconds...`);
    setTimeout(() => makeRequest(currentRetry + 1), RETRY_DELAY);
  } else {
    console.error('Health check failed after all retries');
    process.exit(1); // Failure
  }
}

// Start the health check
makeRequest();
